scipy/sparse/csgraph/_traversal.pyx

```diff
@@ -352,12 +352,22 @@ cpdef breadth_first_order(csgraph, i_start,
         return node_list[:length]
 
 
-cdef unsigned int _breadth_first_directed(
-                           unsigned int head_node,
-                           np.ndarray[ITYPE_t, ndim=1, mode='c'] indices,
-                           np.ndarray[ITYPE_t, ndim=1, mode='c'] indptr,
-                           np.ndarray[ITYPE_t, ndim=1, mode='c'] node_list,
-                           np.ndarray[ITYPE_t, ndim=1, mode='c'] predecessors) noexcept:
+def _breadth_first_directed(
+        unsigned int head_node,
+        np.ndarray[int32_or_int64, ndim=1, mode='c'] indices,
```

Comment:
Memoryviews are preferable over `np.ndarrays`. One can define one as follows:
```suggestion
        int32_or_int64[::1] indices,
```

We can also add `const`-qualification to array which aren't modified (here, CSR matrices'). Yet, there's a bug currently when using it on some custom fused type, see https://github.com/cython/cython/issues/5230.

The other suggestion below proposes a work around of it..

Reply:
How would you feel about deferring the conversion to memoryviews to a followup PR? This module currently uses the ndarray types for everything, so I'd prefer to do a simple maintenance PR where we switch the whole thing over.

We could also handle the const stuff then as well.